### PR TITLE
Fix deadlock in some edge case in autograd

### PR DESCRIPTION
**Pull request by albanD**
Minimal example that deadlocks before but not after:
```python
import torch
from torch.autograd import Function

class Foo(Function):
    @staticmethod
    def forward(ctx, x):
        return x.clone()

    @staticmethod
    def forward(ctx, gO):
        return gO.clone()

def get_out():
    inp = torch.rand(2, requires_grad=True)

    # The python function is first so that it runs
    # last in the backward pass
    right = Foo.apply(inp)

    # An op that creates new memory
    left1 = inp.clone()
    # An op that saves its input
    left2 = left1 ** 2

    # Inplace modify so that the backward for
    # left2 always raises an error
    left1 += 1

    # An op that takes both side as input.
    # After running, both side's last op will be in
    # the ready queue
    # And the op for left will run first as it was
    # executed last during the forward
    out = left2 + right

    return out

# Nothing should be global variables here as, from what
# I can see, python leaks all the global objects
get_out().sum().backward()

```

Since this requires the python interpreter to die, it is hard to test in CI.
Let me know if you have an idea how to do it though.

### CODE REVIEW

File: torch/csrc/autograd/python_function.cpp [hunk 1]
```diff
@@ -167,10 +167,16 @@ auto PyNode::is_traceable() -> bool {
 }
 
 auto PyNode::release_variables() -> void {
-  pybind11::gil_scoped_acquire gil;
-  auto f = (THPFunction*) obj;
-  f->saved_variables.clear();
-  f->has_freed_buffers = 1;
+  // This function is called as part of the Noe destructor!
```
Comment (ezyang) [inline torch/csrc/autograd/python_function.cpp:170]: Noe destructor?

Comment (albanD) [reply torch/csrc/autograd/python_function.cpp:170]: Node*

**Comment by ezyang** [comment]
> Since this requires the python interpreter to die, it is hard to test in CI.
> Let me know if you have an idea how to do it though.

Spawn a fresh Python interpreter for the test. There are a few example of this e.g. test_cublas_config_nondeterministic_alert or `test/test_logging.py`. Make sure timeouts are working correctly.

**Comment by ezyang** [comment]
We should audit the rest of the destructors to check that they don't unconditionally grab the GIL. It is basically never right to unconditionally get the GIL from a destructor.

**Review by ezyang — APPROVED** [review]

**Comment by albanD** [comment]
> We should audit the rest of the destructors to check that they don't unconditionally grab the GIL. It is basically never right to unconditionally get the GIL from a destructor.

That was my idea in the original PR (at least for the autograd object's destructors). But this one is called directly by the custom deleter that we set for each shared_ptr we create that contains a Node. So I completely missed it.

**Pull request closed without merging**